### PR TITLE
ci: Add test coverage job

### LINKED ISSUE #26736: Run coverage functional tests in parallel
**Is your feature request related to a problem? Please describe.**
It currently does not seem possible to execute the functional tests in parallel when running `make cov`.
Thus it makes running this command very slow.

https://github.com/bitcoin/bitcoin/blob/6d40a1a7e7f09ff2c32e53237f968adf8300d028/Makefile.am#L215-L247

**Describe the solution you'd like**
Run test coverage functional tests with `make cov -j8` or `FUNCTIONAL_TESTS_JOBS=8 make cov` ?

### PR DESCRIPTION
**Pull request by aureleoules**
This PR adds a new Cirrus job that generates test coverage reports for all pull requests and master using [Codecov](https://about.codecov.io/). It is free for open-source projects.
We can now quickly determine whether a feature that a pull request adds is properly tested.

I've ran a few tests on my own fork, you can check it out here: https://app.codecov.io/gh/aureleoules/bitcoin and a test pull request: https://app.codecov.io/gh/aureleoules/bitcoin/pull/5.

Codecov has a nice feature that highlights the untested chunks of code directly in the GitHub's "Files changed" tab of a pull request. See for example https://github.com/aureleoules/bitcoin/pull/5/files.

<details>
<summary>Example</summary>

image

</details>

It can also generate a GitHub comment with a brief summary of the coverage: https://github.com/aureleoules/bitcoin/pull/5#issuecomment-1528823591.
Though, that may be undesirable considering Drahtbot already comments on pulls. It can be turned off.

<details>
<summary>Example</summary>

image

</details>

These are the permissions Codecov needs to work:
image

19212b413e518908fa73147b665d8599b7a2f016 also fixes #26736.

### CODE REVIEW

**Comment by aureleoules** [comment]
A backwards compatibility test seems to fail because of BDB, I will fix it if this PR gets Concept ACKed.

**Comment by aureleoules** [comment]
> I think it is good to have coverage reports, but I am not sure if it is good to overlay them in the files tab. It might be too invasive? Maybe just a comment with a link is fine?

You can toggle the annotations in the Files tab client-side using the `a` key.
Or we can fully disable the annotations feature with a custom `codecov.yml` config file but since they are togglable I think it's fine.

It would be nice if Drahtbot could link to the Codecov report also.

> For the other commits: It would be good for reviewers to explain and motivate them, and explain why they are fine to do and what their tradeoffs are (especially not deleting the gcda files?)

I updated the commits descriptions, thanks.

> Also, if the other commits are self-motivated and stand-alone, they can be submitted/reviewed/tested in a separate pull

I think they are small enough to be reviewed in this pull request but let me know if you want me to split them.

File: ci/test/06_script_b.sh [hunk 1]
```diff
@@ -104,3 +104,9 @@ if [ -z "$DANGER_RUN_CI_ON_HOST" ]; then
   echo "Stop and remove CI container by ID"
   docker container kill "${CI_CONTAINER_ID}"
 fi
+
+if [ "$RUN_COVERAGE" = "true" ]; then
+  CI_EXEC make cov
+  CI_EXEC geninfo . -b src -o coverage.info
+  CI_EXEC codecov
```
Comment (dergoegge) [inline ci/test/06_script_b.sh:111]: Maybe add an option to skip this step? then devs could run this job locally, without it failing to upload the reports.

**Review by dergoegge — COMMENTED** [review]
Concept ACK

Thank you for working on this, I think this would be valuable to have.

It seems like it would be easy to replace codecov with something we could host on our own infra (should that become necessary). For now codecov seems fine to me though.

> Read and write access to ... pull requests

Does this mean codecov can push changes to pull request or just that it can comment?

File: ci/test/06_script_b.sh (hunk 1 above)
Comment (aureleoules) [reply ci/test/06_script_b.sh]: Done thanks

**Comment by aureleoules** [comment]
> Does this mean codecov can push changes to pull request or just that it can comment?

It seems that the permissions for "Pull request" are the following: https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#pull-requests

So I don't think Codecov is able to push to PRs.

**Review by maflcko — COMMENTED** [review]
I presume there is a setting to disable the comment and just have DrahtBot add a link in the form of `https://app.codecov.io/gh/{owner}/{repo}/pull/{number}` to the summary comment?

File: .cirrus.yml [hunk 2]
```diff
@@ -366,3 +366,16 @@ task:
   << : *MAIN_TEMPLATE
   env:
     << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
+
+task:
+  name: 'code coverage [lunar]'
```
[the remark this answers is not in the document]
Comment (aureleoules) [reply .cirrus.yml]: Done thanks. I rolled back to jammy.

**Comment by aureleoules** [comment]
> I presume there is a setting to disable the comment and just have DrahtBot add a link in the form of https://app.codecov.io/gh/{owner}/{repo}/pull/{number} to the summary comment?

Yes there is. Done in 320eb03aeb097f6d14757c66a1b0bf40d0473b88.

**Review by pablomartin4btc — COMMENTED** [review]
Concept ACK.

I really like this feature, I find it very useful. Thanks @aureleoules.

I agree with @dergoegge on that perhaps we could replace it with something hosted on us as well and in the meaintime we can take advantage from Codecov.

As [commented](https://github.com/bitcoin/bitcoin/pull/27547#issuecomment-1529712995) by @MarcoFalke on being a bit invasive:
- one thing is the comment on the PR itself, where I can see it's only one line now? I see your example here: https://github.com/aureleoules/bitcoin/pull/5#issuecomment-1528823591.
- on the annotations, on the Files section of the PR, I'm looking at your example here: https://github.com/aureleoules/bitcoin/pull/5/files#annotation_10900357719. I was going to ask you again how to disable/ toggle off the actual comments from Codecov and just found out that you were referring in your [comment](https://github.com/bitcoin/bitcoin/pull/27547#issuecomment-1529730669) actually to the key `A` from the keyboard haha, I like this!

**Comment by aureleoules** [comment]
Thanks for the reviews.
> Alternatively, it may be possible to run this completely outside of CI (locally or on a dedicated machine), which updates the summary comment, once and if the coverage report is ready.

I've actually been working on a self-hosted coverage tool for Bitcoin Core recently. It would run independently of this repo. So I'll close this PR.
I'll probably comment on this PR once I have an MVP ready to be presented.

**Pull request closed without merging**